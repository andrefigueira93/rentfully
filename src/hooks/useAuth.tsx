import { SignUpProps } from "@/schemas/auth";
import { ListUserProps } from "@/schemas/user";
import { authUserAtom, registeredUsersAtom } from "@/store/auth-store";
import bcrypt from "bcryptjs";
import { useAtom } from "jotai";
import React, {
  PropsWithChildren,
  createContext,
  useContext,
  useMemo,
} from "react";

// Define the shape of the context
interface AuthContextData {
  user: Partial<ListUserProps[number]> | null;
  isAuthenticated: boolean;
  signUp: ({ user }: { user: SignUpProps }) => void;
  signIn: ({ email, password }: { email: string; password: string }) => void;
  signOut: () => void;
}
// Create a context with the default value of undefined
const AuthContext = createContext<AuthContextData | undefined>(undefined);

// Create a provider component that will expose the context
const AuthProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [user, setUser] = useAtom(authUserAtom);
  const [registeredUsers, setRegisteredUsers] = useAtom(registeredUsersAtom);

  // Memoed Check if the user is authenticated
  const isAuthenticated = useMemo(() => !!user?.id, [user]);

  // Function to find a user by email
  function findUserByEmail(email: string) {
    return registeredUsers.find((item) => item.email === email);
  }

  // Sign up function to create a new user
  function signUp({ user }: { user: SignUpProps }) {
    const userExists = findUserByEmail(user.email);

    if (userExists?.id) {
      throw new Error("User already exists");
    }

    let newUser = {};

    setRegisteredUsers((prev) => {
      newUser = {
        ...user,
        id: prev.length + 1,
        createdAt: new Date(),
      };
      return [...prev, newUser];
    });

    setUser(newUser);
  }

  // Sign in function to authenticate a user
  function signIn({ email, password }: { email: string; password: string }) {
    const user = findUserByEmail(email);

    if (!user?.id) {
      throw new Error("User not found");
    }

    // Compare the password with the encrypted password
    const decryptedPassword = bcrypt.compareSync(
      password,
      user?.password || ""
    );

    if (!decryptedPassword) {
      throw new Error("Invalid credentials");
    }

    delete user.password;

    setUser(user);
  }

  function signOut() {
    setUser(null);
  }

  return (
    <AuthContext.Provider
      value={{ user, isAuthenticated, signIn, signUp, signOut }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;

export const useAuth = () => {
  const ctx = useContext(AuthContext);
  if (!ctx) throw new Error("`useAuth` must be used within a AuthProvider");
  return ctx;
};
