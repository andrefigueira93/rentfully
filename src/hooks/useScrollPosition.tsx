import React, {
  PropsWithChildren,
  createContext,
  useContext,
  useEffect,
} from "react";

// Define the shape of the context
interface ScrollPositionContextData {
  scrollPosition: {
    x: number;
    y: number;
  };
}

// Create a context with the default value of undefined
const ScrollPositionContext = createContext<
  ScrollPositionContextData | undefined
>(undefined);

// Create a provider component that will expose the context
const ScrollPositionProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [scrollPosition, setScrollPosition] = React.useState({
    x: 0,
    y: 0,
  });

  useEffect(
    () => {
      const handleScroll = () => {
        setScrollPosition({ x: window.scrollX, y: window.scrollY });
      };

      // Add the event listener
      window.addEventListener("scroll", handleScroll);

      // Remove the event listener on cleanup
      return () => {
        window.removeEventListener("scroll", handleScroll);
      };
    },
    [
      // We don't want to run this effect on every render
      // So I'm gonna left this array of dependencies empty
      // https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects
    ]
  );

  return (
    <ScrollPositionContext.Provider value={{ scrollPosition }}>
      {children}
    </ScrollPositionContext.Provider>
  );
};

export default ScrollPositionProvider;

// Create a hook to use the ScrollPositionContext
export const useScrollPosition = () => {
  const ctx = useContext(ScrollPositionContext);
  if (!ctx)
    throw new Error(
      "`useScrollPosition` must be used within a ScrollPositionProvider"
    );
  return ctx;
};
