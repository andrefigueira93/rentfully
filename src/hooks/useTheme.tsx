import { themeAtom } from "@/store/global-store";
import { useAtom } from "jotai";
import { createContext, useContext, useEffect } from "react";

// Theme types
type Theme = "dark" | "light" | "system";

// Define the shape of the props
type ThemeProviderProps = {
  children: React.ReactNode;
};

// Define the shape of the context
type ThemeProviderState = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
};

// Define the initial state of the context
const initialState: ThemeProviderState = {
  theme: "system",
  setTheme: () => null,
};

// Create a context with the default value of undefined
const ThemeProviderContext = createContext<ThemeProviderState>(initialState);

// Create a provider component that will expose the context
export function ThemeProvider({ children, ...props }: ThemeProviderProps) {
  const [theme, setTheme] = useAtom(themeAtom);

  // Set the theme on first load
  useEffect(() => {
    const root = window.document.documentElement;

    root.classList.remove("light", "dark");

    if (theme === "system") {
      const systemTheme = window.matchMedia("(prefers-color-scheme: dark)")
        .matches
        ? "dark"
        : "light";

      root.classList.add(systemTheme);
      return;
    }

    root.classList.add(theme);
  }, [theme]);

  const value = {
    theme,
    setTheme: (theme: Theme) => {
      setTheme(theme);
    },
  };

  return (
    <ThemeProviderContext.Provider {...props} value={value}>
      {children}
    </ThemeProviderContext.Provider>
  );
}

// Create a hook to use the ThemeContext
export const useTheme = () => {
  const context = useContext(ThemeProviderContext);

  if (context === undefined)
    throw new Error("useTheme must be used within a ThemeProvider");

  return context;
};
