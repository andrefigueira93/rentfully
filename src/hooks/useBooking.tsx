import { BookingProps, ListBookingProps } from "@/schemas/booking";
import { bookingFiltersAtom } from "@/store/booking-filters-store";
import { authUserBookingsAtom, bookingsAtom } from "@/store/bookings-store";
import { addDays } from "date-fns";
import { useAtom, useAtomValue } from "jotai";
import React, { PropsWithChildren, createContext, useContext } from "react";
import { useNavigate } from "react-router-dom";
import { toast } from "sonner";
import { useAuth } from "./useAuth";

// Define the shape of the context
interface BookingContextData {
  bookings: Partial<ListBookingProps>;
  myBookings: Partial<ListBookingProps>;
  newBooking: (propertyId: number) => void;
  updateBooking: ({ booking }: { booking: Partial<BookingProps> }) => void;
  deleteBooking: (id: number) => void;
}
// Create a context with the default value of undefined
const BookingContext = createContext<BookingContextData | undefined>(undefined);

// Create a provider component that will expose the context
const BookingProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [bookings, setBookings] = useAtom(bookingsAtom);
  const myBookings = useAtomValue(authUserBookingsAtom);
  const filters = useAtomValue(bookingFiltersAtom);
  const { user } = useAuth();

  const navigate = useNavigate();

  // While we don't have a backend, we will use this function to create a new booking
  // and add it to the bookings array with a fake id and createdAt date
  function newBooking(propertyId: number) {
    if (!user?.id) {
      toast("You should be authenticated to make a booking", {
        action: {
          label: "Sign in",
          onClick: () => navigate("/sign-in"),
        },
        cancel: {
          label: "Nevermind",
          onClick: () => toast.dismiss(),
        },
        icon: <span className="text-2xl">ðŸ˜…</span>,
        important: true,
        position: "top-right",
      });

      if (typeof window !== "undefined") {
        throw new Error("You should be authenticated to make a booking");
      }

      return;
    }

    const today = new Date();

    setBookings((prev) => [
      ...prev,
      {
        id: prev.length + 1,
        propertyId,
        userId: user.id!, // We know that user is not undefined because we checked it above
        status: "Pending",
        checkIn: filters.checkIn ?? today, // If checkIn is not defined, use today
        checkOut: filters.checkOut ?? addDays(today, 1), // If checkOut is not defined, use tomorrow
        guests: filters.numberOfGuests,
        createdAt: today,
      },
    ]);

    navigate("/new-booking");
  }

  // Function to update a booking with updatedAt date
  function updateBooking({ booking }: { booking: Partial<BookingProps> }) {
    setBookings((prev) => {
      return prev.map((item) => {
        if (item.id === booking.id) {
          return {
            ...item,
            ...booking,
            updatedAt: new Date(),
          };
        }
        return item;
      });
    });
  }

  // Function to delete a booking
  function deleteBooking(id: number) {
    setBookings((prev) => prev.filter((item) => item.id !== id));
  }

  return (
    <BookingContext.Provider
      value={{
        bookings,
        myBookings,
        newBooking,
        deleteBooking,
        updateBooking,
      }}
    >
      {children}
    </BookingContext.Provider>
  );
};

export default BookingProvider;

// Create a hook to use the BookingContext
export const useBooking = () => {
  const ctx = useContext(BookingContext);
  if (!ctx)
    throw new Error("`useBooking` must be used within a BookingProvider");
  return ctx;
};
