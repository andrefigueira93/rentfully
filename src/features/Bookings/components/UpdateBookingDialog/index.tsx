import { Button } from "@/components/ui/button";
import { DatePicker } from "@/components/ui/date-picker";
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "@/components/ui/dialog";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { useBooking } from "@/hooks/useBooking";
import { calculateFinalPrice, formatCurrency } from "@/lib/utils";
import { authUserBookingsAtom } from "@/store/bookings-store";
import { differenceInDays } from "date-fns";
import { atom, useAtom, useAtomValue } from "jotai";
import React, { useMemo } from "react";

interface UpdateBookingDialogProps {
  id: number;
}

const checkInDateAtom = atom<Date | undefined>(new Date());
const checkOutDateAtom = atom<Date | undefined>(new Date());
const guestsAtom = atom<number>(1);

const UpdateBookingDialog: React.FC<UpdateBookingDialogProps> = ({ id }) => {
  const bookings = useAtomValue(authUserBookingsAtom);

  const [checkInDate, setCheckInDate] = useAtom(checkInDateAtom);
  const [checkOutDate, setCheckOutDate] = useAtom(checkOutDateAtom);
  const [guests, setGuests] = useAtom(guestsAtom);

  const currentBooking = useMemo(
    () => bookings.find((booking) => booking.id === id),
    [bookings, id]
  );

  const { updateBooking } = useBooking();

  function registerValues() {
    if (!currentBooking) return;
    setCheckInDate(currentBooking.checkIn);
    setCheckOutDate(currentBooking.checkOut);
    setGuests(currentBooking.guests);
  }

  const totalDays = useMemo(() => {
    if (!checkInDate || !checkOutDate) return 1;
    return differenceInDays(checkOutDate, checkInDate);
  }, [checkInDate, checkOutDate]);

  const finalPrice = useMemo(() => {
    return calculateFinalPrice({
      price: currentBooking?.property?.price,
      extraPrice: currentBooking?.property?.extraGuestPrice,
      guests,
      totalDays,
    });
  }, [
    currentBooking?.property?.price,
    currentBooking?.property?.extraGuestPrice,
    guests,
    totalDays,
  ]);

  function handleUpdateBooking() {
    if (!currentBooking) return;
    updateBooking({
      booking: {
        ...currentBooking,
        checkIn: checkInDate,
        checkOut: checkOutDate,
        guests,
        finalPrice,
      },
    });
  }

  return (
    <Dialog onOpenChange={registerValues}>
      <DialogTrigger asChild>
        <Button data-testid="update-button" variant="outline">
          Update
        </Button>
      </DialogTrigger>
      <DialogContent className="sm:max-w-[500px]">
        <DialogHeader>
          <DialogTitle>Edit booking</DialogTitle>
          <DialogDescription>
            Make changes to your booking here. Click save when you're done.
          </DialogDescription>
        </DialogHeader>
        <div className="grid gap-4 py-4">
          <div className="grid grid-cols-4 items-center gap-4">
            <Label htmlFor="name" className="text-right">
              Chek-in Date
            </Label>
            <div className="sm:col-span-3">
              <DatePicker atom={checkInDateAtom} />
            </div>
          </div>
          <div className="grid grid-cols-4 items-center gap-4">
            <Label htmlFor="name" className="text-right">
              Chek-out Date
            </Label>
            <div className="sm:col-span-3">
              <DatePicker atom={checkOutDateAtom} />
            </div>
          </div>
          <div className="grid grid-cols-4 items-center gap-4">
            <Label htmlFor="username" className="text-right">
              Number of Guests
            </Label>
            <div className="sm:col-span-3">
              <Input
                type="number"
                testId="guests-input"
                id="guests"
                name="guests"
                value={guests}
                onChange={(e) => setGuests(parseInt(e.target.value))}
              />
            </div>
          </div>
        </div>
        <DialogFooter className="items-center">
          <strong>New Total: {formatCurrency(finalPrice)}</strong>
          <Button data-testid="save-changes" onClick={handleUpdateBooking}>
            Save changes
          </Button>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
};

export default UpdateBookingDialog;
