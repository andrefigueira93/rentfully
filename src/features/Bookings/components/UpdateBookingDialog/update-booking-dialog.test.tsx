/* eslint-disable @typescript-eslint/no-explicit-any */
import TestWrapper from "@/__tests__/wrapper";
import * as BookingHook from "@/hooks/useBooking";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { addDays, endOfDay, startOfToday } from "date-fns";
import * as Jotai from "jotai";
import { describe, expect, test, vi } from "vitest";
import UpdateBookingDialog from "./index";

vi.mock("jotai", async () => {
  const actualModule = await vi.importActual("jotai");
  return {
    ...(actualModule as any),
    useAtom: vi.fn(),
  };
});

describe("UpdateBookingDialog", () => {
  const today = startOfToday();
  const mockBooking = {
    id: 1,
    propertyId: 1,
    checkIn: today,
    checkOut: endOfDay(addDays(today, 3)),
    guests: 2,
    property: {
      price: 100,
      extraGuestPrice: 20,
    },
  };

  beforeEach(() => {
    vi.spyOn(Jotai, "useAtomValue").mockReturnValue([mockBooking]);
    vi.spyOn(Jotai, "useAtom").mockReturnValue([new Date(), vi.fn() as never]);
  });

  test("renders correctly", () => {
    const { container } = render(<UpdateBookingDialog id={1} />, {
      wrapper: TestWrapper,
    });

    expect(screen.getByTestId("update-button")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("populates initial values when dialog is opened", async () => {
    const { container } = render(<UpdateBookingDialog id={1} />, {
      wrapper: TestWrapper,
    });

    fireEvent.click(screen.getByTestId("update-button"));

    const guestsInput = screen.getByTestId("guests-input");
    expect(guestsInput).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("calls updateBooking on save changes", async () => {
    const mockUpdateBooking = vi.fn();
    vi.spyOn(BookingHook, "useBooking").mockReturnValue({
      bookings: [],
      myBookings: [],
      deleteBooking: vi.fn(),
      updateBooking: mockUpdateBooking,
      newBooking: vi.fn(),
    });

    const { container } = render(<UpdateBookingDialog id={1} />, {
      wrapper: TestWrapper,
    });

    fireEvent.click(screen.getByTestId("update-button"));
    fireEvent.click(screen.getByTestId("save-changes"));

    await waitFor(() => {
      expect(mockUpdateBooking).toHaveBeenCalledWith(
        expect.objectContaining({
          booking: expect.objectContaining({
            id: mockBooking.id,
          }),
        })
      );
    });

    expect(container).toMatchSnapshot();
  });
});
