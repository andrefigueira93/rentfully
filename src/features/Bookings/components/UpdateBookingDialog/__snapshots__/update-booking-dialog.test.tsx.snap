// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpdateBookingDialog > calls updateBooking on save changes 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
>
  <button
    aria-controls="radix-:r8:"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
    data-state="open"
    data-testid="update-button"
    type="button"
  >
    Update
  </button>
</div>
`;

exports[`UpdateBookingDialog > populates initial values when dialog is opened 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
>
  <button
    aria-controls="radix-:r3:"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
    data-state="open"
    data-testid="update-button"
    type="button"
  >
    Update
  </button>
</div>
`;

exports[`UpdateBookingDialog > renders correctly 1`] = `
<div>
  <button
    aria-controls="radix-:r0:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
    data-state="closed"
    data-testid="update-button"
    type="button"
  >
    Update
  </button>
</div>
`;
